🏗 Add dev-mode for `amp visual-diff` and split code into modules

danielrozenberg opened the pull request:
This PR adds a "development mode" for `amp visual-diff` to help iterate on new visual diff tests faster. It's triggered by calling `amp visual-diff --dev` (preferably with `--grep <test-name>` to make it easier):

![image](https://user-images.githubusercontent.com/1839738/139736902-360f6ec8-8401-4064-ae8a-351ae09bc9b7.png)

Then the user selects the test file and an interactive test if they exist:

![image](https://user-images.githubusercontent.com/1839738/139737003-b3ec2d2e-42db-4347-9a48-3e8ad6755c82.png)

And the task opens a browser window with the test, runs any interactivity code embedded in it, and then prompts the user for actions:

![image](https://user-images.githubusercontent.com/1839738/139737244-5c60328c-4927-4fba-ab7d-7016812fb5d1.png)

The actions are to immediately test whether a CSS selector is visible/invisible on the page or to reload the page.

![image](https://user-images.githubusercontent.com/1839738/139737597-66ab67b3-f16b-4ff3-9e79-0a823a94e838.png)

Reloading the page by pressing enter on an empty prompt will reload any new changes to the HTML and interactive JS files

rileyajones commented on build-system/tasks/visual-diff/dev-mode.js:
```diff
@@ -0,0 +1,177 @@
+ */
+async function devMode(browser, webpages) {
+  /** @type {WebpageDef} */
+  let webpage;
+  if (webpages.length > 1) {
+    webpage = (
+      await inquirer.prompt([
+        {
+          type: 'list',
+          name: 'webpage',
+          message:
+            'Select test name from ' +
+            cyan('visual-diff.jsonc') +
+            ' (use ' +
+            cyan('--grep') +
+            ' to filter this list):',
+          choices: webpages.map((webpage) => ({
+            name: webpage.name,
+            value: webpage,
+          })),
+        },
+      ])
+    ).webpage;
+  } else {
+    webpage = webpages[0];
+    log(
+      'info',
+      'Using test',
+      yellow(webpage.name),
+      '(Only one test matches the',
+      cyan('--grep'),
+      'value)'
+    );
+  }
```
Breaking this out into a helper function could help with the readability and allow `webpage` to be `const` and always defined.

danielrozenberg replied:
Good idea! Done.

rsimha commented on build-system/tasks/visual-diff/browser.js:
```diff
@@ -0,0 +1,196 @@
+// 5. Go to "Tools" -> "Version information"
+// 6. Paste the full version (add ".0" at the end) in the "Version" field and click "Lookup"
+// 7. Copy the value next to "Branch Base Position" and update the line below
+const PUPPETEER_CHROMIUM_REVISION = '870763'; // 91.0.4472.0
```
Not directly related to this PR: I'm concerned that at some point, pinning to v91 will prevent us from detecting bugs on newer versions of Chrome. Is there any value in running the tests on the latest version even if they are eventually rendered on v91 by Percy?

danielrozenberg replied:
Not really, this is just about the rendering. Percy does update its engine a few times a year, so we can always keep track of that

rsimha commented on build-system/tasks/visual-diff/dev-mode.js:
```diff
@@ -0,0 +1,200 @@
+
+const inquirer = require('inquirer');
+const path = require('path');
+const puppeteer = require('puppeteer'); // eslint-disable-line no-unused-vars
```
Since this is purely for typing, any chance it can be replaced by something inside https://www.npmjs.com/package/@types/puppeteer?

(The disable directive is not an actual concern, so feel free to ignore.)

danielrozenberg replied:
This throws `Cannot import type declaration files. Consider importing 'puppeteer' instead of '@types/puppeteer'.ts(6137)`

Another option is to use the `import(...).Type` syntax which I personally find unreadable:
```js
/**
 * @param {!import('puppeteer').Browser} browser a Puppeteer controlled browser.
 */
```

lmk if you have a strong opinion

rsimha replied:
Cool, leaving this as is SGTM.

rsimha reviewed (approved):
This is excellent, thank you for putting it together. LGTM with a couple minor non-blocking comments.

danielrozenberg commented:
@rsimha @rileyajones I updated the instructions for writing visual diff tests. It became a bit too much to be included in testing.md so I split it into a separate file. PTAL?

rsimha commented on docs/writing-visual-diff-tests.md:
```diff
@@ -0,0 +1,55 @@
+
+### Validate the test
+
+Finally, verify that your test display as expected by executing it on Percy:
```
nit: display -> displays

danielrozenberg replied:
Done.

rsimha commented on docs/writing-visual-diff-tests.md:
````diff
@@ -0,0 +1,55 @@
+```
+
+-   When the test finishes executing it will print a URL to Percy where you can inspect the results. It should take about a minute to finish processing.
+-   Inspect the build on Percy. If you are not happy with the results, fix your page or code, and repeat. If all is well, approve it. This creates a new baseline on Percy, against which all following builds will be compared.
````
suggestion: Move "if all is well..." to the next step for clarity.

danielrozenberg replied:
Done.

rsimha reviewed (approved):
Docs LGTM.
